File: backend/src/export_report.py
from datetime import datetime
from io import BytesIO

import pandas as pd
from openpyxl.styles import Font, PatternFill
from openpyxl.utils import get_column_letter

# (sheet title, results key, columns to keep in that order)
PHASES = [
    ("Meter Coverage", "step0", ["Match Key", "Value", "Client Name", "Issue"]),
    ("Metadata", "step1", ["Meter Number", "Client Name", "Mismatched Field", "Original Field (Hebrew)", "Alteco Value", "Client Value"]),
    ("Consumption", "step2", ["Meter Number", "Client Name", "Mismatched Field", "Original Field (Hebrew)", "Alteco Value", "Client Value"]),
    ("Financial", "step3", ["Customer ID", "Client Name", "Mismatched Field", "Original Field (Hebrew)", "Alteco Value", "Client Value"]),
]

HEADER_FILL = PatternFill(start_color="1F2937", end_color="1F2937", fill_type="solid")
HEADER_FONT = Font(bold=True, color="FFFFFF")


def _style_sheet(ws, ncols):
    for col_idx in range(1, ncols + 1):
        cell = ws.cell(row=1, column=col_idx)
        cell.font = HEADER_FONT
        cell.fill = HEADER_FILL
    ws.freeze_panes = "A2"
    ws.auto_filter.ref = ws.dimensions
    for col_idx in range(1, ncols + 1):
        letter = get_column_letter(col_idx)
        max_len = max(
            (len(str(ws.cell(row=r, column=col_idx).value or "")) for r in range(1, ws.max_row + 1)),
            default=10,
        )
        ws.column_dimensions[letter].width = min(max(max_len + 2, 12), 60)


def build_discrepancy_workbook(results):
    """
    Builds a formatted .xlsx report from a reconciliation results dict
    ({"step0": [...], "step1": [...], "step2": [...], "step3": [...]}).
    Sheet order: Summary first, then one sheet per phase (bold header row,
    frozen header, auto-filter, auto-sized columns) — empty phases still get
    a sheet, with a clear "no discrepancies" note instead of being skipped.
    Returns a BytesIO positioned at 0, ready to stream as a response body.
    """
    buffer = BytesIO()
    with pd.ExcelWriter(buffer, engine="openpyxl") as writer:
        summary_rows = [
            {"Phase": title, "Mismatches": len(results.get(key) or [])}
            for title, key, _cols in PHASES
        ]
        df_summary = pd.DataFrame(summary_rows)
        df_summary.to_excel(writer, sheet_name="Summary", index=False)
        writer.sheets["Summary"]["C1"] = "Generated"
        writer.sheets["Summary"]["C2"] = datetime.now().strftime("%Y-%m-%d %H:%M")
        _style_sheet(writer.sheets["Summary"], len(df_summary.columns))

        for title, key, columns in PHASES:
            rows = results.get(key) or []
            if rows:
                df = pd.DataFrame(rows)
                df = df[[c for c in columns if c in df.columns]]
            else:
                df = pd.DataFrame([{"Result": "No discrepancies found"}])
            df.to_excel(writer, sheet_name=title, index=False)
            _style_sheet(writer.sheets[title], len(df.columns))

    buffer.seek(0)
    return buffer

File: backend-ts/src/exportReport.ts
import ExcelJS from 'exceljs';

export type PhaseKey = 'step0' | 'step1' | 'step2' | 'step3';

interface PhaseDef {
  title: string;
  key: PhaseKey;
  columns: string[];
}

// (sheet title, results key, columns to keep in that order)
const PHASES: PhaseDef[] = [
  { title: 'Meter Coverage', key: 'step0', columns: ['Match Key', 'Value', 'Client Name', 'Issue'] },
  {
    title: 'Metadata',
    key: 'step1',
    columns: ['Meter Number', 'Client Name', 'Mismatched Field', 'Original Field (Hebrew)', 'Alteco Value', 'Client Value'],
  },
  {
    title: 'Consumption',
    key: 'step2',
    columns: ['Meter Number', 'Client Name', 'Mismatched Field', 'Original Field (Hebrew)', 'Alteco Value', 'Client Value'],
  },
  {
    title: 'Financial',
    key: 'step3',
    columns: ['Customer ID', 'Client Name', 'Mismatched Field', 'Original Field (Hebrew)', 'Alteco Value', 'Client Value'],
  },
];

const HEADER_FILL: ExcelJS.Fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF1F2937' } };
const HEADER_FONT: Partial<ExcelJS.Font> = { bold: true, color: { argb: 'FFFFFFFF' } };

function styleSheet(worksheet: ExcelJS.Worksheet, ncols: number): void {
  const headerRow = worksheet.getRow(1);
  for (let col = 1; col <= ncols; col++) {
    const cell = headerRow.getCell(col);
    cell.font = HEADER_FONT;
    cell.fill = HEADER_FILL;
  }
  worksheet.views = [{ state: 'frozen', ySplit: 1 }];
  worksheet.autoFilter = { from: { row: 1, column: 1 }, to: { row: 1, column: ncols } };

  for (let col = 1; col <= ncols; col++) {
    let maxLen = 10;
    for (let row = 1; row <= worksheet.rowCount; row++) {
      const value = worksheet.getRow(row).getCell(col).value;
      const len = value === null || value === undefined ? 0 : String(value).length;
      if (len > maxLen) maxLen = len;
    }
    worksheet.getColumn(col).width = Math.min(Math.max(maxLen + 2, 12), 60);
  }
}

/**
 * Builds a formatted .xlsx report from a reconciliation results object
 * ({step0: [...], step1: [...], step2: [...], step3: [...]}). Sheet order:
 * Summary first, then one sheet per phase (bold header row, frozen header,
 * auto-filter, auto-sized columns) -- empty phases still get a sheet, with a
 * clear "no discrepancies" note instead of being skipped. Returns a Buffer
 * ready to stream as a response body.
 */
export async function buildDiscrepancyWorkbook(results: Record<string, Array<Record<string, unknown>>>): Promise<Buffer> {
  const workbook = new ExcelJS.Workbook();

  const summarySheet = workbook.addWorksheet('Summary');
  summarySheet.addRow(['Phase', 'Mismatches']);
  for (const { title, key } of PHASES) {
    summarySheet.addRow([title, (results[key] ?? []).length]);
  }
  summarySheet.getCell('C1').value = 'Generated';
  summarySheet.getCell('C2').value = formatTimestamp(new Date());
  styleSheet(summarySheet, 2);

  for (const { title, key, columns } of PHASES) {
    const rows = results[key] ?? [];
    const worksheet = workbook.addWorksheet(title);
    if (rows.length > 0) {
      const presentColumns = columns.filter((c) => Object.prototype.hasOwnProperty.call(rows[0], c));
      worksheet.addRow(presentColumns);
      for (const row of rows) {
        worksheet.addRow(presentColumns.map((c) => row[c] ?? null));
      }
      styleSheet(worksheet, presentColumns.length);
    } else {
      worksheet.addRow(['Result']);
      worksheet.addRow(['No discrepancies found']);
      styleSheet(worksheet, 1);
    }
  }

  const arrayBuffer = await workbook.xlsx.writeBuffer();
  return Buffer.from(arrayBuffer);
}

function formatTimestamp(date: Date): string {
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())} ${pad(date.getHours())}:${pad(date.getMinutes())}`;
}
